components/sync15/src/migrate_state.rs

```diff
@@ -0,0 +1,199 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::record_types::MetaGlobalRecord;
+use crate::state::PersistedGlobalState;
+use crate::CollSyncIds;
+use serde_json::Value;
+
+/// Given a string persisted as our old GlobalState V1 struct, extract out
+/// the sync IDs for the collection, plus a string which should be used as the
+/// new "global persisted state" (which holds the declined engines).
+/// Returns (None, None) in early error cases (eg, invalid JSON, wrong schema
+/// version etc). Otherwise, you can expect the returned global state to be
+/// Some, even if the CollSyncIds is None (which can happen if the engine is
+/// missing, or flagged for reset)
+pub fn extract_v1_state(
+    state: Option<String>,
+    collection: &'static str,
+) -> (Option<CollSyncIds>, Option<String>) {
```

Comment:
I wonder if this should be an enum. I guess it's probably fine.

Reply:
Yeah, I don't think this is too confusing, so I left this alone.